### PR TITLE
fixes #22411 - unattended: render template macros from plugins

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

File: app/controllers/unattended_controller.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 class UnattendedController < ApplicationController
-  include Foreman::Renderer
+  include UnattendedHelper # also includes Foreman::Renderer
```
Comment (ohadlevy) [inline app/controllers/unattended_controller.rb]: are we actually using it to begin with? I'm sure there are some dead methods in there?

Comment (timogoebel) [reply app/controllers/unattended_controller.rb]: I have no clue, the whole rendering scope should not be the controller that renders the template imho.

Actually, I still have another issue with this:

```ruby
# fresh foreman-rake console
UnattendedController.ancestors # does not include ForemanOmaha::RendererMethods
UnattendedController.send(:include, UnattendedHelper)
UnattendedController.ancestors # includes ForemanOmaha::RendererMethods now
```

**Comment by ohadlevy** [comment]
> Actually, I still have another issue with this:

to clarify - issue with the PR or with the current (pre pr) situation?

I'm OK with merging if that's what you mean.

**Comment by timogoebel** [comment]
> to clarify - issue with the PR or with the current (pre pr) situation?

I don't think, this PR solves my issue. But I need to do some further investigation.

**Comment by timogoebel** [comment]
This now contains a working fix for my issue.

**Review by lzap — APPROVED** [review]
This works for me. Our templating engine needs an overhaul. It's complicated, we have three contexts, two modes (safe and nonsafe). One day, we will hopefully have just one safe renderer used everywhere.

File: app/registries/foreman/plugin.rb [hunk 2]
```diff
@@ -440,7 +440,7 @@ def permission_table_exists?
     def extend_template_helpers_by_module(mod)
       mod = mod.constantize
 
-      (TemplatesController.descendants + [ TemplatesController, UnattendedHelper ]).each do |klass|
+      (TemplatesController.descendants + [ TemplatesController, UnattendedController, UnattendedHelper ]).each do |klass|
```
[the remark this answers is not in the document]
Comment (lzap) [reply app/registries/foreman/plugin.rb]: Can you fix that.

**Review by lzap — CHANGES_REQUESTED** [review]
Actually can you extend `test_extend_rendering_helpers` test to verify new behavior?

**Comment by timogoebel** [comment]
> Our templating engine needs an overhaul.

I hope, @ares and I make some time on the construction day to refactor it. If we can reduce the contexts to one, that'd be a great benefit already. Actually out test_helper in community templates os in a better shape than foreman core. That's a little embarrassing.

**Comment by lzap** [comment]
I struggle to understand why it does not work, we include `UnattendedHelper` in `UnattendedController` so why the method is not there. Maybe because it's too late right? What I mean by that:

* UnattendedController includes UnattendedHelper
* plugin mechanism includes into UnattendedHelper

**Comment by timogoebel** [comment]
> I struggle to understand why it does not work

@lzap: Me too. Already spent hours digging into that. This is the best I could come up with so far. It's not much, but I'm at a point where "rewriting the whole thing" might be faster 😕

**Comment by lzap** [comment]
I think it's the ordering what matters, look:

```ruby
>> module A; def a; end end
=> :a

>> module B; def b; end end
=> :b

>> class X; include A; end
=> X

>> X.new.a
=> nil

>> A.send(:include, B)
=> A

>> X.new.a
=> nil

>> X.new.b
NoMethodError: undefined method `b' for #<X:0x005617d6006e48>

>> class X; include A; end
=> X

>> X.new.b
=> nil
```

In other words, class `X` needs to "reinclude" to get access to extra methods included via send (plugin engine).

**Comment by timogoebel** [comment]
Finding out the reason for the "method missing" is actually pretty easy. If I add a debugger and log `self.class.ancestors` in `UnattendedController#safe_render`, without this patch I get this:

```
[ForemanBootdisk::UnattendedControllerExt, UnattendedController, UnattendedHelper, Foreman::Renderer, Foreman::ForemanUrlRenderer, #<Module:0x00007ffb4b92fee8>, #<Module:0x00007ffb4a0e3150>, ApplicationController, ApplicationShared, FindCommon, Foreman::ThreadSession::Cleaner, Foreman::Controller::TopbarSweeper, Foreman::Controller::Session, Foreman::Controller::Authentication, Foreman::Controller::MigrationChecker, #<Module:0x00007ffb487daa50>, #<Module:0x00007ffb446d21e8>, #<Module:0x00007ffb446d2210>, ActionController::Base, Turbolinks::XHRHeaders, Turbolinks::Cookies, Turbolinks::XDomainBlocker, Turbolinks::Redirection, SecureHeaders, ActiveRecord::Railties::ControllerRuntime, ActionDispatch::Routing::RouteSet::MountedHelpers, ActionController::RespondWith, ActionController::ParamsWrapper, ActionController::Instrumentation, ActionController::Rescue, ActionController::HttpAuthentication::Token::ControllerMethods, ActionController::HttpAuthentication::Digest::ControllerMethods, ActionController::HttpAuthentication::Basic::ControllerMethods, ActionController::DataStreaming, ActionController::Streaming, ActionController::ForceSSL, ActionController::RequestForgeryProtection, AbstractController::Callbacks, ActiveSupport::Callbacks, ActionController::FormBuilder, ActionController::Flash, ActionController::Cookies, ActionController::ParameterEncoding, ActionController::StrongParameters, ActiveSupport::Rescuable, ActionController::ImplicitRender, ActionController::BasicImplicitRender, ActionController::MimeResponds, AbstractController::Caching, AbstractController::Caching::ConfigMethods, AbstractController::Caching::Fragments, ActionController::Caching, ActionController::EtagWithFlash, ActionController::EtagWithTemplateDigest, ActionController::ConditionalGet, ActionController::Head, ActionController::Renderers::All, ActionController::Renderers, ActionController::Rendering, ActionView::Layouts, ActionView::Rendering, ActionController::Redirecting, ActiveSupport::Benchmarkable, AbstractController::Logger, ActionController::UrlFor, AbstractController::UrlFor, ActionDispatch::Routing::UrlFor, ActionDispatch::Routing::PolymorphicRoutes, ActionController::Helpers, AbstractController::Helpers, AbstractController::AssetPaths, AbstractController::Translation, AbstractController::Rendering, ActionView::ViewPaths, ActionController::Metal, ActionController::Testing::Functional, AbstractController::Base, ActiveSupport::Configurable, ActiveSupport::ToJsonWithActiveSupportEncoder, Object, Foreman::Util, Foreman::Gettext::Debug, FastGettext::TranslationMultidomain, PP::ObjectMixin, GettextI18nRails::HtmlSafeTranslations, FastGettext::Translation, FriendlyId::ObjectUtils, ActiveSupport::Dependencies::Loadable, JSON::Ext::Generator::GeneratorMethods::Object, ActiveSupport::Tryable, Kernel, BasicObject]
```

With this patch, it looks like this:
```
[ForemanBootdisk::UnattendedControllerExt, UnattendedController, ForemanOmaha::RendererMethods, ForemanBootdisk::RendererMethods, UnattendedHelper, Foreman::Renderer, Foreman::ForemanUrlRenderer, #<Module:0x00007fd2adcf1430>, #<Module:0x00007fd2a9eb5220>, ApplicationController, ApplicationShared, FindCommon, Foreman::ThreadSession::Cleaner, Foreman::Controller::TopbarSweeper, Foreman::Controller::Session, Foreman::Controller::Authentication, Foreman::Controller::MigrationChecker, #<Module:0x00007fd2abdbb368>, #<Module:0x00007fd2ab9d1dd8>, #<Module:0x00007fd2ab9d1e28>, ActionController::Base, Turbolinks::XHRHeaders, Turbolinks::Cookies, Turbolinks::XDomainBlocker, Turbolinks::Redirection, SecureHeaders, ActiveRecord::Railties::ControllerRuntime, ActionDispatch::Routing::RouteSet::MountedHelpers, ActionController::RespondWith, ActionController::ParamsWrapper, ActionController::Instrumentation, ActionController::Rescue, ActionController::HttpAuthentication::Token::ControllerMethods, ActionController::HttpAuthentication::Digest::ControllerMethods, ActionController::HttpAuthentication::Basic::ControllerMethods, ActionController::DataStreaming, ActionController::Streaming, ActionController::ForceSSL, ActionController::RequestForgeryProtection, AbstractController::Callbacks, ActiveSupport::Callbacks, ActionController::FormBuilder, ActionController::Flash, ActionController::Cookies, ActionController::ParameterEncoding, ActionController::StrongParameters, ActiveSupport::Rescuable, ActionController::ImplicitRender, ActionController::BasicImplicitRender, ActionController::MimeResponds, AbstractController::Caching, AbstractController::Caching::ConfigMethods, AbstractController::Caching::Fragments, ActionController::Caching, ActionController::EtagWithFlash, ActionController::EtagWithTemplateDigest, ActionController::ConditionalGet, ActionController::Head, ActionController::Renderers::All, ActionController::Renderers, ActionController::Rendering, ActionView::Layouts, ActionView::Rendering, ActionController::Redirecting, ActiveSupport::Benchmarkable, AbstractController::Logger, ActionController::UrlFor, AbstractController::UrlFor, ActionDispatch::Routing::UrlFor, ActionDispatch::Routing::PolymorphicRoutes, ActionController::Helpers, AbstractController::Helpers, AbstractController::AssetPaths, AbstractController::Translation, AbstractController::Rendering, ActionView::ViewPaths, ActionController::Metal, ActionController::Testing::Functional, AbstractController::Base, ActiveSupport::Configurable, ActiveSupport::ToJsonWithActiveSupportEncoder, Object, Foreman::Util, Foreman::Gettext::Debug, FastGettext::TranslationMultidomain, PP::ObjectMixin, GettextI18nRails::HtmlSafeTranslations, FastGettext::Translation, FriendlyId::ObjectUtils, ActiveSupport::Dependencies::Loadable, JSON::Ext::Generator::GeneratorMethods::Object, ActiveSupport::Tryable, Kernel, BasicObject]
```
Note, that `ForemanOmaha::RendererMethods` and others are missing without this patch. It must be a timing issue. I'll dig a little more.

**Comment by timogoebel** [comment]
@lzap: Maybe we can add something like this:

https://github.com/theforeman/foreman/blob/66e834071e015ed76cd679f2486dc549016ffba7/config/application.rb#L220-L227

Not, that it's particularly pretty, but if it works...

**Comment by lzap** [comment]
I lean towards current proposal, I like it. Can you just fix the test? Perhaps instead of testing method presence in individual classes, can you simply squash together all methods from all descendants of both controllers and search in there? This will test actual behavior.

**Comment by timogoebel** [comment]
@lzap: I fixed the tests. The timing issue can come from plugins that extend "UnattendedController" (like bootdisk) before this code runs.

> can you simply squash together all methods from all descendants of both controllers and search in there

I don't follow.

**Review by lzap — APPROVED** [review]
It's fine as it is currently tested. Thanks.

File: app/controllers/unattended_controller.rb [hunk 3]
```diff
@@ -249,7 +249,7 @@ def safe_render(template)
     end
 
     begin
-      render :inline => "<%= unattended_render(@unsafe_template_content, @template_name).html_safe %>"
+      render :plain => unattended_render(@unsafe_template_content, @template_name)
```
Comment (dLobatog) [inline app/controllers/unattended_controller.rb:252]: This breaks my provisioning, are you able to get a kickstart template with this change in @timogoebel @lzap ?

Comment (lzap) [reply app/controllers/unattended_controller.rb:252]: Yeah filed https://github.com/theforeman/foreman/pull/5258 let's discuss there.

**Pull request merged**